src/java/org/streamingpool/core/service/streamfactory/FlattenedStreamFactory.java

```diff
@@ -42,18 +41,23 @@
  * @see FlattenedStreamId
  */
 public class FlattenedStreamFactory implements StreamFactory {
-
     @Override
-    public <T> Optional<Publisher<T>> create(StreamId<T> id, DiscoveryService discoveryService) {
+    public <T> ErrorStreamPair<T> create(StreamId<T> id, DiscoveryService discoveryService) {
         if (!(id instanceof FlattenedStreamId)) {
-            return empty();
+            return ErrorStreamPair.empty();
         }
-        FlattenedStreamId<T> filteredId = (FlattenedStreamId<T>) id;
-        StreamId<Iterable<T>> source = filteredId.sourceStreamId();
 
-        return Optional.of(Flowable.fromPublisher(discoveryService.discover(source))
-                .filter(Objects::nonNull)
-                .flatMap(Flowable::fromIterable)
-                .filter(Objects::nonNull));
+        FlattenedStreamId<T> flattenedStreamId = (FlattenedStreamId<T>) id;
+        return createFlattenedStream(flattenedStreamId, discoveryService);
+    }
+
+    private <T> ErrorStreamPair<T> createFlattenedStream(FlattenedStreamId<T> id, DiscoveryService discoveryService) {
+        Flowable<Iterable<T>> sourceStream = Flowable.fromPublisher(discoveryService.discover(id.sourceStreamId()));
+        ErrorDeflector ed = ErrorDeflector.create();
+
+        return ed.stream(sourceStream.flatMap(iterable -> {
```

Comment:
I think you do not really need the error deflector here (you use only the stream method, which in your case is more or less equivalent to ErrorStreamPair.ofData(....)...

If you want you could wrap the next line into an ed.emptyOnException() and then use ed.streamNonEmpty().
However, I agree that I hardly see a case where your conversion might fail ....